src/Tools/dotnet-monitor/Exceptions/ConfiguredExceptionsStore.cs

```diff
@@ -0,0 +1,138 @@
+        IExceptionsStore,
+        IAsyncDisposable
+    {
+        private readonly ExceptionsStore _store;
+
+        public ConfiguredExceptionsStore(IOptions<ExceptionsOptions> options)
+            : this(options.Value.GetTopLevelLimit(), callback: null)
+        {
+        }
+
+        internal ConfiguredExceptionsStore(int topLevelLimit, ExceptionsStoreCallback callback = null)
+        {
+            if (topLevelLimit <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(topLevelLimit));
+            }
+
+            _store = new ExceptionsStore(new Callback(this, topLevelLimit, callback));
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            return _store.DisposeAsync();
+        }
+
+        public void AddExceptionInstance(IExceptionsNameCache cache, ulong exceptionId, ulong groupId, string message, DateTime timestamp, ulong[] stackFrameIds, int threadId, ulong[] innerExceptionIds, string activityId, ActivityIdFormat activityIdFormat)
+        {
+            _store.AddExceptionInstance(cache, exceptionId, groupId, message, timestamp, stackFrameIds, threadId, innerExceptionIds, activityId, activityIdFormat);
+        }
+
+        public void RemoveExceptionInstance(ulong exceptionId)
+        {
+            _store.RemoveExceptionInstance(exceptionId);
+        }
+
+        public IReadOnlyList<IExceptionInstance> GetSnapshot()
+        {
+            return _store.GetSnapshot();
+        }
+
+        private sealed class Callback : ExceptionsStoreCallback
+        {
+            private readonly ExceptionsStoreCallback _callback;
+            private readonly Dictionary<ulong, List<ulong>> _innerExceptionMap = new();
+            private readonly Dictionary<ulong, List<ulong>> _outerExceptionMap = new();
+            private readonly ConfiguredExceptionsStore _store;
+            private readonly LinkedList<ulong> _topLevelExceptions = new();
+            private readonly int _topLevelLimit;
+
+            public Callback(ConfiguredExceptionsStore store, int topLevelLimit, ExceptionsStoreCallback callback)
+            {
+                _callback = callback;
+                _store = store;
+                _topLevelLimit = topLevelLimit;
+            }
+
+            public override void AfterAdd(IExceptionInstance instance)
+            {
+                _callback?.AfterAdd(instance);
+            }
+
+            public override void BeforeAdd(IExceptionInstance instance)
+            {
+                _outerExceptionMap.Add(instance.Id, new List<ulong>());
+                _innerExceptionMap.Add(instance.Id, instance.InnerExceptionIds.ToList());
+
+                // Make sure all of the inner exceptions are no longer top-level exceptions.
+                foreach (ulong innerExceptionId in instance.InnerExceptionIds)
+                {
+                    if (_outerExceptionMap.TryGetValue(innerExceptionId, out List<ulong> outerExceptions))
```

Comment:
Might be my own fault for not looking at this long enough, but maybe add a comment for this logic - I think I'm not fully grasping the use of the `_outerExceptionMap`.

Okay - I think I get it - for efficiency (traversing) reasons this gives us bidirectionality so we can reference whether an exception is an inner exception of anything else (i.e. who are all the parents of this particular exception). I'm keeping the first part of this comment in since I still think more explanation might be helpful for anyone skimming this file.

Reply:
Added more comments describing what an outer exception is and that it's easier/faster to track the information this way rather than having to find IExceptionInstances.